### ISSUE TITLE
[dynamo] Implement iter fallback as polyfill

### ISSUE THREAD

**Issue by jon-chuang** [pytorch/pytorch#112727]
### 🚀 The feature, motivation and pitch

It should be fairly simple to implement iter fallback as a polyfill

The non-general thing we do now is to handle repeat as a special case.

### Alternatives

Not handle general iterators

**Comment by jon-chuang** [comment]
@ezyang, come to think of it, should we just get rid of all the iterator special casing and use polyfills for everything by relying on the underlying objects? lol

**Comment by jon-chuang** [comment]
We'd get rid of all the special-case code implemented in https://github.com/pytorch/pytorch/pull/110967 https://github.com/pytorch/pytorch/pull/110953, as well as the special case `ListIteratorVariable`...

The main reason we want to track `ListIteratorVariable` is actually mutation, so this needs to be figured out - how can we make polyfills compatible with mutation, and with general iterators?

**Comment by jon-chuang** [comment]
Another obstacle to this is properly handling `RAISE_VARARGS`. This is due to most iterators' reliance on `StopIteration` to terminate the iterator.

**Comment by jon-chuang** [comment]
Here is some sample code to show how this would work:

```python
def next_variable(self):
    try:
      ..
    except StopIteration:
      return ConstantVariable(StopIteration)

def builtin_next():
    next_var, next_iter = iter.next_variable()
    if next_var.is_python_constant() and next_var.as_python_constant() == StopIteration:
        raise StopIteration()  # or unimplemented("raising StopIteration exception within dynamo")

def FOR_ITER:
    next_var, next_iter = iter.next_variable()
    if next_var.is_python_constant() and next_var.as_python_constant() == StopIteration:
        self.jump(inst)

def YIELD_FROM(self, inst):
    next_var, next_iter = iter.next_variable()
    if next_var.is_python_constant() and next_var.as_python_constant() == StopIteration:
        return
```

**Comment by ezyang** [comment]
Yeah we kind of don't have any exception support in Dynamo, so absent a pressing use case, it seems better to do that first, and then use the normal exception mechanism to handle StopIteration.

**Comment by anijain2305** [comment]
cc @XuehaiPan in case you already fixed this. We have both polyfill and exception handling in Dynamo now.
